### PR TITLE
upgrade webpack from 1 to 2

### PR DESCRIPTION
**Pull request by rauleite**
Migration completed, as follows:
https://webpack.js.org/guides/migrating/

why?
https://medium.com/webpack/webpack-2-and-beyond-40520af9067f

### CODE REVIEW

**Comment by rauleite** [comment]
Migration completed, as follows:
https://webpack.js.org/guides/migrating/

**Comment by NathanCH** [comment]
Any thoughts on passing sassLoader options directly instead of through `LoaderOptionsPlugin`? The implementation in this PR will break style paths when used components.

ie.

    {
      test: /\.scss$/,
      use: [
        {
          loader: 'sass-loader',
          options: {
            sourceMap: true,
            includePaths: [
              project.paths.client('styles')
            ]
          }
        }
      ]
    }

More info: https://github.com/webpack-contrib/sass-loader/issues/285#issuecomment-278010902

**Comment by rauleite** [comment]
Hi Nathan, I'm sorry I can not help you right now because of lack of time =/

I hope someone helps you!

**Comment by NathanCH** [comment]
No worries man, and thanks. I can confirm this is working, I'll make a separate PR if it doesn't make it into the webpack upgrade.

**Comment by rauleite** [comment]
Very good! Good luck with this.

**Comment by Slumber86** [comment]
I think you forgot an option for sass-loader

```      options : {
        sourceMap : true,
        includePaths: [...project.paths.client('styles'), './node_modules',]
      }```

**Comment by Slumber86** [comment]
Also you missed to migrate from babel to webpack the ES2015 module resolution:
in project.config.js
```
  compiler_babel : {
    cacheDirectory : true,
    plugins        : ['transform-runtime'],
    presets        : [["es2015", { "modules": false }], 'react', 'stage-0']
  },
```

**Comment by collinwu** [comment]
Any status on this? I have an upcoming project and I'd love to leverage this repo, but with webpack 2 :)

**Comment by piu130** [comment]
Why don't we switch to [neutrino](https://github.com/mozilla-neutrino/neutrino-dev)? It's much easier and you don't have to debug build stuff anymore. See [#1177](https://github.com/davezuko/react-redux-starter-kit/issues/1177).

**Comment by EricThsi** [comment]
It's works follow your codes, thx ;) @rauleite

File: config/karma.config.js [hunk 1]
```diff
@@ -23,6 +23,7 @@ const karmaConfig = {
   browsers : ['PhantomJS'],
   webpack  : {
     devtool : 'cheap-module-source-map',
+    entry: './tests/test-bundler.js',
```
Comment (whawker) [inline config/karma.config.js]: Should this be `./${project.dir_test}/test-bundler.js` like on [line 11](https://github.com/davezuko/react-redux-starter-kit/pull/1189/files#diff-c3ce07ad1a971575599220c4a08f29f0R11) of this file

**Comment by prestonbernstein** [comment]
What's the status on this?

File: config/webpack.config.js [hunk 2]
```diff
@@ -84,21 +87,22 @@ if (__TEST__ && !argv.watch) {
...
   webpackConfig.plugins.push(
     new webpack.HotModuleReplacementPlugin(),
-    new webpack.NoErrorsPlugin()
+    new webpack.NoEmitOnErrorsPlugin()
   )
 } else if (__PROD__) {
   debug('Enabling plugins for production (OccurenceOrder, Dedupe & UglifyJS).')
   webpackConfig.plugins.push(
     new webpack.optimize.OccurrenceOrderPlugin(),
     new webpack.optimize.DedupePlugin(),
```
Comment (whawker) [inline config/webpack.config.js]: Probably should remove the DedupePlugin? https://webpack.js.org/guides/migrating/#dedupeplugin-has-been-removed

File: config/webpack.config.js [hunk 3]
```diff
@@ -198,17 +286,22 @@ webpackConfig.module.loaders.push(
...
-  webpackConfig.module.loaders.filter((loader) =>
-    loader.loaders && loader.loaders.find((name) => /css/.test(name.split('?')[0]))
-  ).forEach((loader) => {
-    const first = loader.loaders[0]
-    const rest = loader.loaders.slice(1)
-    loader.loader = ExtractTextPlugin.extract(first, rest.join('!'))
-    delete loader.loaders
+  webpackConfig.module.rules.filter(rule =>
+    rule.loader && /css/.test(rule.loader)
+  ).forEach(loader => {
```
Comment (dineshvgp) [inline config/webpack.config.js]: I think, this should be `rule` instead of `loader`

File: config/webpack.config.js [hunk 4]
```diff
@@ -118,75 +122,159 @@ if (!__TEST__) {
...
+      sassLoader : {
+        includePaths : project.paths.client('styles')
+      }
+    }
   })
-]
+)
 
 // File loaders
 /* eslint-disable */
```
Comment (matgessel) [inline config/webpack.config.js]: can eliminate eslint bypass?

Comment (matgessel) [inline config/webpack.config.js]: Less boilerplate:

    const FONT_EXTS = Map([
      'woff', 'application/font-woff',
      'woff2', 'application/font-woff2',
      'otf', 'font/opentype',
      'ttf', 'application/octet-stream',
      'eot', 'application/vnd.ms-fontobject',
      'svg', 'image/svg+xml',
    ]);
    for (let [extension, mimeType] of FONT_EXTS) {
      webpackConfig.module.rules.push({
        test    : `\\.${extension}(\\?.*)?$`,
        loader  : 'url-loader',
        options : {
          prefix   : 'fonts/',
          name     : '[path][name].[ext]',
          limit    : '10000',
          mimetype : mimeType,
        },
      });
    };
    webpackConfig.module.rules.push(
      {
        test    : /\.(png|jpg|gif)$/,
        loader  : 'url-loader',
        options : {
          limit    : '8192',
        },
      },
    );

(Note addition of .gif)
I haven't tested this code yet.

File: config/webpack.config.js [hunk 5]
```diff
@@ -118,75 +122,159 @@ if (!__TEST__) {
...
   test    : /\.(js|jsx)$/,
   exclude : /node_modules/,
-  loader  : 'babel',
+  loader  : 'babel-loader',
   query   : project.compiler_babel
 }, {
   test   : /\.json$/,
-  loader : 'json'
+  loader : 'json-loader'
 }]
```
Comment (matgessel) [inline config/webpack.config.js]: Can eliminate this. JSON loader is built-in in Webpack 2.

File: config/webpack.config.js [hunk 6]
```diff
@@ -118,75 +122,159 @@ if (!__TEST__) {
...
-    },
-    discardUnused : false,
-    mergeIdents   : false,
-    reduceIdents  : false,
-    safe          : true,
-    sourcemap     : true
+webpackConfig.plugins.push(
+  new webpack.LoaderOptionsPlugin({
+    options : {
+      postcss : [
```
Comment (prestonbernstein) [inline config/webpack.config.js]: add a context of '/' right after options to solve issues when using fonts.

```js
new webpack.LoaderOptionsPlugin({
  options: {
    context: '/',
    ...
  }
})
```

File: config/webpack.config.js (hunk 3 above)
Comment (freele) [inline config/webpack.config.js]: isn't it a misspell here?
```
...
  ).forEach(rule => {
...
```

**Comment by davezuko** [comment]
Moving to https://github.com/davezuko/react-redux-starter-kit/pull/1243.

Thanks for your work, I'll try to cherry-pick some of your fixes over. I have private project that just recently bumped webpack to 2.0, so I'm currently porting all of those changes over here.

**Pull request closed without merging**